### PR TITLE
add aten::upsample_linear1d aten::upsample_linear1d_backward

### PR DESCRIPTION
**Pull request by huaiyuzh**

### CODE REVIEW

File: src/ATen/native/xpu/UpSampleLinear1d.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
+#include <ATen/ATen.h>
+#include <ATen/native/xpu/sycl/UpSampleLinear1dKernels.h>
+#include <ATen/xpu/XPUNativeFunctions.h>
+#include <comm/RegisterUtils.h>
+#include "ATen/core/ATen_fwd.h"
+
+namespace at {
+
+static C10_UNUSED std::array<int64_t, 3> upsample_1d_common_check(
```
Comment (majing921201) [inline src/ATen/native/xpu/UpSampleLinear1d.cpp:9]: move to UpSample.h

File: src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    std::optional<double> scales,
+    Tensor& output) {
+  int output_width = output_size[0];
+  output.zero_();
+  int input_width = input.size(2);
+
+  AT_ASSERT(input_width > 0 && output_width > 0);
+
+  const int num_kernels = output_width;
+  const int num_threads = 512;
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:94]: can we replace the hard code work group size with max_group_size ?

File: src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    bool align_corners,
+    std::optional<double> scales,
+    Tensor& grad_input) {
+  int output_width = output_size[0];
+  int input_width = input_size[2];
+  Tensor grad_output = grad_output_.contiguous();
+  grad_input.zero_();
+
+  const int num_kernels = output_width;
+  const int num_threads = 512;
```
Comment (majing921201) [inline src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:196]: ditto

File: src/ATen/native/xpu/UpSampleLinear1d.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    const Tensor& grad_output,
+    IntArrayRef output_size,
+    IntArrayRef input_size,
+    bool align_corners,
+    std::optional<double> scales,
+    Tensor& grad_input) {
+  upsample_linear1d_backward_meta(
+      grad_output, output_size, input_size, align_corners, scales, grad_input);
+
+  globalContext().alertNotDeterministic("upsample_linear1d_backward_out_xpu");
```
Comment (majing921201) [inline src/ATen/native/xpu/UpSampleLinear1d.cpp:128]: move it to kernel, it is for kernel implementation

File: src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        auto odata = output.packed_accessor64<scalar_t, 3>();
+
+        using accscalar_t = at::acc_type_device<scalar_t, kXPU>;
+        const accscalar_t rwidth = area_pixel_compute_scale<accscalar_t>(
+            input_width, output_width, align_corners, scales);
+        const int num_kernels = output_width;
+        using KernelClass =
+            UpsampleLinear1dKernelFunctor<scalar_t, accscalar_t>;
+        int64_t local_range = syclMaxWorkGroupSize<KernelClass>();
+        KernelClass kfn(num_kernels, rwidth, align_corners, idata, odata);
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:108]: Suggest using `syclMaxWorkGroupSize(kfn)` not forward-declaring `KernelClass`, if ctor of kernel functor class doesn't depend on work group size.

File: src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+};
+
+void upsample_linear1d_backward_kernel(
+    const Tensor& grad_output_,
+    IntArrayRef output_size,
+    IntArrayRef input_size,
+    bool align_corners,
+    std::optional<double> scales,
+    Tensor& grad_input) {
+  globalContext().alertNotDeterministic("upsample_linear1d_backward_out_xpu");
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:188]: Does CUDA has deterministic implementation?

Comment (fengyuan14) [reply src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:188]: `upsample_linear1d_backward_xpu`

File: src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+  int output_width = output_size[0];
+  int input_width = input_size[2];
+  Tensor grad_output = grad_output_.contiguous();
+  grad_input.zero_();
+
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      grad_output.scalar_type(),
+      "upsample_linear1d_backward",
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:199]: `upsample_linear1d_backward_xpu`

File: src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        using accscalar_t = at::acc_type_device<scalar_t, kXPU>;
+        const int num_kernels = output_width;
+        auto idata = grad_input.packed_accessor64<scalar_t, 3>();
+        auto odata = grad_output.packed_accessor64<const scalar_t, 3>();
+        const accscalar_t rwidth = area_pixel_compute_scale<accscalar_t>(
+            input_width, output_width, align_corners, scales);
+        using KernelClass =
+            UpsampleLinear1dBackwardKernelFunctor<scalar_t, accscalar_t>;
+        int64_t local_range = syclMaxWorkGroupSize<KernelClass>();
+        KernelClass kfn(num_kernels, rwidth, align_corners, idata, odata);
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:210]: Ditto.

File: src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+            UpsampleLinear1dKernelFunctor<scalar_t, accscalar_t>;
+        int64_t local_range = syclMaxWorkGroupSize<KernelClass>();
+        KernelClass kfn(num_kernels, rwidth, align_corners, idata, odata);
+        auto global_range = (num_kernels + local_range - 1) / local_range;
+        sycl_kernel_submit(
+            global_range * local_range,
+            local_range,
+            getCurrentSYCLQueue(),
+            kfn);
+      });
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:114]: `global_range` means total work items. So,
```
auto global_range = ((num_kernels + local_range - 1) / local_range) * local_range;
sycl_kernel_submit(
            global_range,
            local_range,
            getCurrentSYCLQueue(),
            kfn);

```

File: src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        const int num_kernels = output_width;
+        auto idata = grad_input.packed_accessor64<scalar_t, 3>();
+        auto odata = grad_output.packed_accessor64<const scalar_t, 3>();
+        const accscalar_t rwidth = area_pixel_compute_scale<accscalar_t>(
+            input_width, output_width, align_corners, scales);
+        using KernelClass =
+            UpsampleLinear1dBackwardKernelFunctor<scalar_t, accscalar_t>;
+        int64_t local_range = syclMaxWorkGroupSize<KernelClass>();
+        KernelClass kfn(num_kernels, rwidth, align_corners, idata, odata);
+        auto global_range = (num_kernels + local_range - 1) / local_range;
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:209]: Ditto.

File: src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp (hunk 6 above)
Comment (huaiyuzh) [reply src/ATen/native/xpu/sycl/UpSampleLinear1dKernels.cpp:187]: CUDA is not deterministic.

**Review by fengyuan14 — APPROVED** [review]

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**